### PR TITLE
JSON serialization controlled by a JSON flag

### PR DESCRIPTION
**Pull request by mfine**
Lay out a JSON serialization path next to the Pickle serialization path. This will line-based serialize out SBP in JSON with the following format:

```python
    {'delta': delta, 'timestamp': timestamp, 'data': data} : dict
      delta = msec reference timestamp (int)
      timestamp = current timestamp (int - UTC epoch)
      data = Python object to JSON serialize
```

Keeping the *meta* info (delta, timestamp) out of the object getting serialized, but open to other opinions.

/cc @fnoble @mookerji

### CODE REVIEW

**Comment by fnoble** [comment]
looks good to me, @mookerji any comment before I merge?

**Comment by mfine** [comment]
@fnoble @mookerji any comments on the JSON block per line? It's making the loading in the analysis a little more complicated. Note, output looks like:

```json
{"timestamp": 1425585366, "data": {"sender": 1497, "msg_type": 256, "crc": 51307, "length": 11, "preamble": 85, "payload": "Kgecv+AY+2///wA="}, "delta": 117251}
{"timestamp": 1425585366, "data": {"sender": 1497, "msg_type": 513, "crc": 32287, "length": 34, "preamble": 85, "payload": "nL/gGA6PwWaY4UJAFbLDFOeYXsA7Di4UIZEAwAAAAAAIAA=="}, "delta": 117252}
{"timestamp": 1425585366, "data": {"sender": 1497, "msg_type": 512, "crc": 28484, "length": 32, "preamble": 85, "payload": "nL/gGELTXq/6oUTBJCdwlkhDUMGunqCHLKNNQQAACAA="}, "delta": 117252}
{"timestamp": 1425585366, "data": {"sender": 1497, "msg_type": 517, "crc": 13058, "length": 22, "preamble": 85, "payload": "nL/gGPD///8GAAAA1v///wAAAAAIAA=="}, "delta": 117252}
{"timestamp": 1425585366, "data": {"sender": 1497, "msg_type": 516, "crc": 45398, "length": 20, "preamble": 85, "payload": "nL/gGO7////Z////DQAAAAAACAA="}, "delta": 117252}
{"timestamp": 1425585366, "data": {"sender": 1497, "msg_type": 21, "crc": 28510, "length": 13, "preamble": 85, "payload": "AACqQQCAeEM57UrFGQ=="}, "delta": 117316}
{"timestamp": 1425585366, "data": {"sender": 1497, "msg_type": 16, "crc": 37598, "length": 37, "preamble": 85, "payload": "YWNxOiBQUk4gMjYgZm91bmQgQCAtMzI0NiBIeiwgMjEgU05SCg=="}, "delta": 117316}
{"timestamp": 1425585366, "data": {"sender": 1497, "msg_type": 16, "crc": 52030, "length": 20, "preamble": 85, "payload": "Tm8gY2hhbm5lbHMgZnJlZSA6KAo="}, "delta": 117316}
{"timestamp": 1425585366, "data": {"sender": 1497, "msg_type": 256, "crc": 58228, "length": 11, "preamble": 85, "payload": "KgcAwOAYYVwAAAA="}, "delta": 117348}
```

**Comment by fnoble** [comment]
Is there any typical convention that people use with JSON? I imagine you have the same problem when streaming JSON in a web context. Do people typically use a block per line there too?

**Comment by mfine** [comment]
Convention I've used in the past is JSON dictionary per line, but then you have to parse the JSON per line vs the entire file. Some relevant links:

http://stackoverflow.com/questions/10699953/format-for-writing-a-json-log-file
http://blog.nodejs.org/2012/03/28/service-logging-in-json-with-bunyan/

**Comment by fnoble** [comment]
I'm sold, especially seeing as there looks to be existing tooling we can use that follows this convention.

**Comment by mookerji** [comment]
Sweet. 

The dict-per-line thing looks just fine. Give me a second to look the rest of it over...

File: scripts/serial_link.py [hunk 1]
```diff
@@ -334,6 +360,36 @@ def format_log_entry(t0, item):
...
+  Returns
+  ----------
+  {'delta': delta, 'timestamp': timestamp, 'data': data} : dict
+    delta = msec reference timestamp (int)
+    timestamp = current timestamp (int - UTC epoch)
+    data = Python object to JSON serialize
+  """
+  timestamp = calendar.timegm(time.gmtime())
+  delta = int((time.time() - t0)*1000)
+  data = {
```
Comment (mookerji) [inline scripts/serial_link.py]: With this here, I actually think that the json serialization logic should live with the `SBP` object definition in `scripts/sbp_piksi.py`, either as an `@property; def json(self): ...` or `def to_json(self): ...`.

Comment (mfine) [inline scripts/serial_link.py]: Agree - I had an iteration done that way... but its more like returning a dict that will be used to JSON serialize - like `to_json_dict` or something - and then doing the actual serialization in the formatter (and adding the extra elements to the envelope).

Comment (mookerji) [inline scripts/serial_link.py]: Ah yeah, sorry: that's actually closer to what I meant (I should have been more specific). The creation of the dict from the SBP attributes is separate concern from the timestamped JSON record that you're emitting. I think `libsbp` will also need some kind of corresponding `to_json_dict`/`from_json_dict` methods for reifying SBP objects.

**Comment by mfine** [comment]
@mookerji Added a `to_json_dict` method. Going to add a `from_json_dict` to SBP objects in the log analysis, too.

**Comment by fnoble** [comment]
Agree about the `to_json_dict` method being on the SBP class. This looks great!

**Pull request merged**